http/basic/src/main/java/io/cloudevents/http/HttpMessageFactory.java

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright 2018-Present The CloudEvents Authors
+ * <p>
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * <p>
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * <p>
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.cloudevents.http;
+
+import io.cloudevents.CloudEvent;
+import io.cloudevents.core.format.EventFormat;
+import io.cloudevents.core.message.MessageReader;
+import io.cloudevents.core.message.MessageWriter;
+import io.cloudevents.core.message.impl.GenericStructuredMessageReader;
+import io.cloudevents.core.message.impl.MessageUtils;
+import io.cloudevents.core.message.impl.UnknownEncodingMessageReader;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+
+public class HttpMessageFactory {
+
+    public static MessageReader createReader(Map<String,String> headers, byte[] body) throws IllegalArgumentException {
```

Comment:
@slinkydeveloper @eamonnmcmanus what about changing this to`Map<String, List<String>>`?

Reply:
That's technically more correct, since HTTP headers can be repeated. It would also allow you to pass in JAX-RS's `MultivaluedMap` directly. On the other hand, the CloudEvents spec doesn't use any repeated headers, and it's a bit less convenient to synthesize a `Map<String, List<String>>` in tests. But I think either type is defensible.

Reply:
Well `java.net.HttpURLConnection` and `com.sun.net.httpserver.HttpServer` also use `Map<String, List<String>>`.

Reply:
I'm against this, because in the cloudevents spec the assumption is that there is one value for each key, so that doesn't really map to the cloudevents spec itself (although, for http, it makes sense)

Reply:
Maybe we can have an override that takes `Map<String, List<String>>` and translates it to `Map<String, String>`

Reply:
That's a fair point. Tests can always use a trivial helper method that makes a `Map<String, List<String>>` out of a `Map<String, String>`:

```
private static Map<String, List<String>> toMultimap(Map<String, String> map) {
  return map.entrySet().stream().collect(toMap(e -> e.getKey(), e -> List.of(e.getValue())));
}
```

I'd suggest having both via overloads, but erasure would get in the way. So `Map<String, List<String>>` is probably better if you have to choose.

Reply:
Would method overload via generics even work in Java? Or do you mean it would have slightly different name?

Reply:
You can encapsulate in a `Wire` instance (name tbd) the actual headers and payload, then to create the wire you have methods like:

```
Wire of(Map<String, String>, byte[])
Wire ofMultimap(Map<String, String>, byte[])
```